src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs

```diff
@@ -19,88 +17,112 @@ namespace Microsoft.CodeAnalysis;
 {
-    internal async Task<LinkedFileMergeSessionResult> MergeDiffsAsync(IMergeConflictHandler mergeConflictHandler, CancellationToken cancellationToken)
+    internal async Task<LinkedFileMergeSessionResult> MergeDiffsAsync(IMergeConflictHandler? mergeConflictHandler, CancellationToken cancellationToken)
     {
         var sessionInfo = new LinkedFileDiffMergingSessionInfo();
 
-        var linkedDocumentGroupsWithChanges = solutionChanges
-            .GetProjectChanges()
-            .SelectMany(p => p.GetChangedDocuments())
-            .GroupBy(d => oldSolution.GetDocument(d).FilePath, StringComparer.OrdinalIgnoreCase);
+        using var _ = PooledDictionary<string, List<(Document newDocument, ImmutableArray<byte> newContentHash)>>.GetInstance(out var filePathToNewDocumentsAndHashes);
+        foreach (var documentId in solutionChanges.GetProjectChanges().SelectMany(p => p.GetChangedDocuments()))
+        {
+            var newDocument = newSolution.GetRequiredDocument(documentId);
+            var filePath = newDocument.FilePath;
+            Contract.ThrowIfNull(filePath);
 
-        var linkedFileMergeResults = new List<LinkedFileMergeResult>();
+            if (!filePathToNewDocumentsAndHashes.TryGetValue(filePath, out var newDocumentsAndHashes))
+            {
+                newDocumentsAndHashes = [];
+                filePathToNewDocumentsAndHashes.Add(filePath, newDocumentsAndHashes);
+            }
+
+            var newText = await newDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+            var newContentHash = newText.GetContentHash();
+            // Ignore any linked documents that we have the same contents as.  
+            if (newDocumentsAndHashes.Any(t => t.newContentHash.SequenceEqual(newContentHash)))
+                continue;
+
+            newDocumentsAndHashes.Add((newDocument, newContentHash));
+        }
 
         var updatedSolution = newSolution;
-        foreach (var linkedDocumentsWithChanges in linkedDocumentGroupsWithChanges)
+        var linkedFileMergeResults = new List<LinkedFileMergeResult>();
+
+        foreach (var (filePath, newDocumentsAndHashes) in filePathToNewDocumentsAndHashes)
         {
-            var documentInNewSolution = newSolution.GetDocument(linkedDocumentsWithChanges.First());
+            Contract.ThrowIfTrue(newDocumentsAndHashes.Count == 0);
 
-            // Ensure the first document in the group is the first in the list of 
-            var allLinkedDocuments = documentInNewSolution.GetLinkedDocumentIds().Add(documentInNewSolution.Id);
-            if (allLinkedDocuments.Length == 1)
-            {
+            // Don't need to do anything if this document has no linked siblings.
+            var firstNewDocument = newDocumentsAndHashes[0].newDocument;
+            var relatedDocuments = newSolution.GetRelatedDocumentIds(firstNewDocument.Id);
+            if (relatedDocuments.Length == 1)
```

Comment:
this is the next meat part.  if we've only truly come up with one change to a doc, then we don't even have to do any diff/merge work.  we can just take that change and apply to all linked files.

Reply:
note: could potentially lift this up to the dictionary building loop.  i will consider that.

Reply:
i don't see the GetContentHash being expensive at all (it's 0.2% of hte total trace).  so preventing this by lifting this to the dictionary loop doesn't really help.